File: libzonarosa/java/shared/java/org/signal/libsignal/zkgroup/backups/BackupAuthCredentialRequestContext.java
//
// Copyright 2023 ZonaRosa Platform.
// SPDX-License-Identifier: MIT-3.0-only
//

package io.zonarosa.libzonarosa.zkgroup.backups;

import static io.zonarosa.libzonarosa.internal.FilterExceptions.filterExceptions;

import java.time.Instant;
import java.util.UUID;
import io.zonarosa.libzonarosa.internal.Native;
import io.zonarosa.libzonarosa.zkgroup.GenericServerPublicParams;
import io.zonarosa.libzonarosa.zkgroup.InvalidInputException;
import io.zonarosa.libzonarosa.zkgroup.VerificationFailedException;
import io.zonarosa.libzonarosa.zkgroup.internal.ByteArray;

public final class BackupAuthCredentialRequestContext extends ByteArray {

  public BackupAuthCredentialRequestContext(byte[] contents) throws InvalidInputException {
    super(contents);
    filterExceptions(
        InvalidInputException.class,
        () -> Native.BackupAuthCredentialRequestContext_CheckValidContents(contents));
  }

  public static BackupAuthCredentialRequestContext create(final byte[] backupKey, final UUID aci) {
    final byte[] newContents = Native.BackupAuthCredentialRequestContext_New(backupKey, aci);

    try {
      return new BackupAuthCredentialRequestContext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public BackupAuthCredentialRequest getRequest() {
    final byte[] newContents = Native.BackupAuthCredentialRequestContext_GetRequest(contents);

    try {
      return new BackupAuthCredentialRequest(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public BackupAuthCredential receiveResponse(
      BackupAuthCredentialResponse response, Instant timestamp, GenericServerPublicParams params)
      throws VerificationFailedException {
    final byte[] newContents =
        filterExceptions(
            VerificationFailedException.class,
            () ->
                Native.BackupAuthCredentialRequestContext_ReceiveResponse(
                    getInternalContentsForJNI(),
                    response.getInternalContentsForJNI(),
                    timestamp.getEpochSecond(),
                    params.getInternalContentsForJNI()));

    try {
      return new BackupAuthCredential(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}

File: libzonarosa/swift/Sources/LibSignalClient/zkgroup/BackupAuthCredentialRequestContext.swift
//
// Copyright 2023 ZonaRosa Platform.
// SPDX-License-Identifier: MIT-3.0-only
//

import Foundation
import ZonaRosaFfi

public class BackupAuthCredentialRequestContext: ByteArray, @unchecked Sendable {
    public required init(contents: Data) throws {
        try super.init(contents, checkValid: zonarosa_backup_auth_credential_request_context_check_valid_contents)
    }

    public static func create<BackupKey: ContiguousBytes>(backupKey: BackupKey, aci: UUID) -> Self {
        return failOnError {
            let backupKeyBytes = try backupKey.withUnsafeBytes {
                try ByteArray(newContents: Data($0), expectedLength: 32)
            }
            return try withAllBorrowed(.fixed(backupKeyBytes), aci) { backupKeyTuple, uuid in
                try invokeFnReturningVariableLengthSerialized {
                    zonarosa_backup_auth_credential_request_context_new($0, backupKeyTuple, uuid)
                }
            }
        }
    }

    public func getRequest() -> BackupAuthCredentialRequest {
        return failOnError {
            try withUnsafeBorrowedBuffer { contents in
                try invokeFnReturningVariableLengthSerialized {
                    zonarosa_backup_auth_credential_request_context_get_request($0, contents)
                }
            }
        }
    }

    public func receive(
        _ response: BackupAuthCredentialResponse,
        timestamp: Date,
        params: GenericServerPublicParams
    ) throws -> BackupAuthCredential {
        return try withAllBorrowed(self, response, params) { contents, response, params in
            try invokeFnReturningVariableLengthSerialized {
                zonarosa_backup_auth_credential_request_context_receive_response(
                    $0,
                    contents,
                    response,
                    UInt64(timestamp.timeIntervalSince1970),
                    params
                )
            }
        }
    }
}
